libraries/AP_Math/quaternion.cpp

```diff
@@ -424,34 +424,47 @@ void Quaternion::from_vector312(float roll, float pitch, float yaw)
 }
 
 // create a quaternion from its axis-angle representation
-void Quaternion::from_axis_angle(Vector3f v)
+void Quaternion::from_axis_angle(const Vector3f &v)
 {
-    const float theta = v.length();
-    if (is_zero(theta)) {
-        q1 = 1.0f;
-        q2=q3=q4=0.0f;
-        return;
+    const float theta_sq = v.x * v.x + v.y * v.y + v.z * v.z;
```

Comment:
You could use this here instead?
```suggestion
    const float theta_sq = v.length_squared();
```

Reply:
Will do, wasn't aware that this method existed.

Reply:
There are a lot of those little functions around to clean things up. 

> Tracked down the issue --- there are times when a Quaternion is set to [0, 0, 0, 0], which causes a divide by zero in the `to_axis_angle()` function. Edited my code to protect against that, now all checks passed.
> 
> What's worrying me now is that somewhere in the codebase a zero Quaternion, which is not a valid rotation representation, may be created. Neither does the `normalize()` member protect against that situation.



> Did someone leave, then retract, a comment about my using the `f` literal suffix on numbers and using more double precision? (saw the gmail notification) The Quaternion class has float data members so the `f` suffix makes sense. I'd glad to rewrite the Quaternion class as a template and do another PR, pls let me know.



> Did someone leave, then retract, a comment about my using the `f` literal suffix on numbers and using more double precision? (saw the gmail notification) The Quaternion class has float data members so the `f` suffix makes sense. I'd glad to rewrite the Quaternion class as a template and do another PR, pls let me know if I should try that out.

Sorry! I accidentally hit submit review before I meant to as that [0,0,0,0] quaternion you got should be figured out.  That is totally preexisting and not for this PR.